odoo/tools/mail.py

```diff
@@ -177,82 +105,168 @@ def parse_style(self, el):
-        cleaned = u'<p>ParserError when sanitizing</p>'
-    except Exception:
-        if not silent:
-            raise
-        logger.warning(u'unknown error obtained when sanitizing %r', src, exc_info=True)
-        cleaned = u'<p>Unknown error when sanitizing</p>'
-
-    # this is ugly, but lxml/etree tostring want to put everything in a 'div' that breaks the editor -> remove that
-    if cleaned.startswith(u'<div>') and cleaned.endswith(u'</div>'):
-        cleaned = cleaned[5:-6]
-
-    return markupsafe.Markup(cleaned)
+        raise
+
+    # perform quote detection before cleaning and class removal
+    if doc is not None:
+        for el in doc.iter(tag=etree.Element):
+            tag_quote(el)
+
+    if callback:
+        doc = callback(doc)
+
+    src = html.tostring(doc, encoding='unicode')
+
+    # this is ugly, but lxml/etree tostring want to put everything in a
+    # 'div' that breaks the editor -> remove that
+    if src.startswith(u'<div>') and src.endswith(u'</div>'):
+        src = src[5:-6]
+
+    return src
+
+
+def html_sanitize(src, silent=True, sanitize_tags=True, sanitize_attributes=False, sanitize_style=False, sanitize_form=True, strip_style=False, strip_classes=False):
+    if not src:
+        return src
+
+    logger = logging.getLogger(__name__ + '.html_sanitize')
+
+    def sanitize_handler(doc):
+        kwargs = {
+            'page_structure': True,
+            'style': strip_style,              # True = remove style tags/attrs
+            'sanitize_style': sanitize_style,  # True = sanitize styling
+            'forms': sanitize_form,            # True = remove form tags
+            'remove_unknown_tags': False,
+            'comments': False,
+            'processing_instructions': False
+        }
+        if sanitize_tags:
+            kwargs.update(SANITIZE_TAGS)
+
+        if sanitize_attributes:  # We keep all attributes in order to keep "style"
+            if strip_classes:
+                current_safe_attrs = safe_attrs - frozenset(['class'])
+            else:
+                current_safe_attrs = safe_attrs
+            kwargs.update({
+                'safe_attrs_only': True,
+                'safe_attrs': current_safe_attrs,
+            })
+        else:
+            kwargs.update({
+                'safe_attrs_only': False,  # keep oe-data attributes + style
+                'strip_classes': strip_classes,  # remove classes, even when keeping other attributes
+            })
+
+        try:
+            # some corner cases make the parser crash, such as:
+            # - <SCRIPT/XSS SRC=\"http://ha.ckers.org/xss.js\"></SCRIPT> in test_mail
+            # - HTML comment only string like `<!-- abc -->`
+            # - whitespace only string
+            cleaner = _Cleaner(**kwargs)
+            cleaned = cleaner.clean_html(doc)
+        except etree.ParserError as e:
+            if 'empty' in str(e):
+                return u""
```

Comment:
:+1:

I also wonder if we should really keep the bare `Exception` case, it seems to have been introduced alongside those commits:
- [d7185be](https://github.com/odoo/odoo/commit/d7185be38c2d386c1826d7e0aefacf15828d7050#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR57) (not yet)
- [a74d50d](https://github.com/odoo/odoo/commit/a74d50d220b26f26a37faeb59e8923f1e635c2ea#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR55) (not yet, double exception added because of retrying)
- [f97bd8b](https://github.com/odoo/odoo/commit/f97bd8bd6313db276452df1e4be176fa0d2ce292#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR65) (here)
- [6fd552c](https://github.com/odoo/odoo/commit/6fd552c3ceae47384aca1ac960cb3fd5f8bbb832#diff-b583c3495a9c5bf0e224788074ddaabf14647a6e2b8cca1fe273492986b187acR91) (fine tuned)

I guess it's anyway not worth the risk in 16.0